External extension

jadams-tresys opened the pull request:
There are still a couple circularly deadlocked issues preventing round tripping on a 4 tests.  Wanted to get things pushed up for review to both take a look at that and get the ball rolling for getting these changes pushed up.

These changes allow for the Raster schemas to override the types defined in nitf_ude_types.dfdl.xsd to replace the hexBinary sections with Raster specific elements.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -165,36 +182,58 @@ The message root is 'NITF'.
+                  <xs:complexType>
+                    <xs:sequence>
+                      <xs:element name="OriginatorsName" type="nitf:ECS-A" dfdl:length="27" nillable="true" dfdl:ref="nitf:nilString" />
+                      <xs:element name="OriginatorsPhoneNumber" type="nitf:ECS-A" dfdl:length="18" nillable="true" dfdl:ref="nitf:nilString" />
```
Looks like OriginatorsPhoneNumer can e moved out of the choice?

Also, I wonder if there's a benefit to avoiding the HeaderInfo elements? FileBackGroundColor could be made optional with occursCount depending on the nitif version. And OriginatosrsName could have a length expression depending on the version as well. Might make the infoset a little cleaner. I prefer to avoid version specific elements when possible. That way if someone ever needs to access a field (e.g. OriginatorsName) they don't need to know the version to figure out which HeaderInfo to look into.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -624,68 +664,7 @@ The message root is 'NITF'.
-                  </xs:sequence>
-                </xs:complexType> 
-              </xs:element>
+              <xs:element name="Data" type="ude:image_data_type" dfdl:lengthKind="explicit" dfdl:length="{ ../../Header/ImageSegmentLengths[dfdl:occursIndex()]/DataLength }" />
```
The rest of this file uses camel case for types. For consistency, should we rename this ImageDataType?

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -45,17 +45,23 @@ The message root is 'NITF'.
            xmlns:fn="http://www.w3.org/2005/xpath-functions"
            xmlns:daf="urn:ogf:dfdl:2013:imp:daffodil.apache.org:2018:ext"
            xmlns:nitf="urn:nitf:2.1"
+           xmlns:ude="urn:nitfUserDefinedExtensions"
```
Thoughts on a prefix like "nitfx" or something? Makes it more clear that this is nitf related? Or maybe even something different? To me, the word "extensions" almost implies an it's optional, but anyone using NITF must create the types for the schema to work. Not sure what a better name would be.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -853,50 +833,72 @@ The message root is 'NITF'.
-            </xs:sequence>
-          </xs:complexType>
-        </xs:element>
+        <xs:element name="TaggedRecordExtension" type="ude:tre_type" minOccurs="0" maxOccurs="unbounded" dfdl:occursCountKind="parsed" />
```
Same question as above, should we make camel case?

stevedlawrence commented on src/test/resources/com/tresys/nitf/xsd/nitf_ude_types.dfdl.xsd:
```diff
@@ -0,0 +1,156 @@
+  AND NONINFRINGEMENT. IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE 
+  FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT 
+  OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
+  OR OTHER DEALINGS WITH THE SOFTWARE. -->
```
This copyright header got mangled.

stevedlawrence commented on src/test/resources/com/tresys/nitf/xsd/nitf_ude_types.dfdl.xsd:
```diff
@@ -0,0 +1,156 @@
+<?xml version='1.0' encoding='UTF-8'?>
```
I wonder if this we should split this repo into two sub projects, something like nitf and nitf-core? nitf-core would have the current nitif.dfdl.xsd file? And then nitf would contain this file and all the nitf tests? That way, nitf can depend on nitf-core and work the way it does now, and variations like raster can depend on nitf-core and implement the necessary extension types for added functionality?

stevedlawrence commented on src/test/resources/com/tresys/nitf/xsd/nitf_ude_types.dfdl.xsd:
```diff
@@ -0,0 +1,156 @@
+  xmlns:nitf="urn:nitf:2.1"
+  elementFormDefault="unqualified">
+
+  <xs:import namespace="urn:nitf:2.1" schemaLocation="com/tresys/nitf/xsd/nitf.dfdl.xsd" />
```
Is this import needed to use the simple types defined in nitf? Maybe we should split those types into a nitf-types-types.dfld.xsd or something like that? That way there's a clear separation between the main nitf schema and the types that extensions can use?

jadams-tresys replied:
I actually had done that earlier when trying to get past the class path issues.  I think it makes sense to make the change and I like the idea of splitting up the NITF project into 2 separate projects.

stevedlawrence commented on src/test/resources/com/tresys/nitf/xsd/nitf_ude_types.dfdl.xsd:
```diff
@@ -0,0 +1,156 @@
+        truncateSpecifiedLengthString="no" utf16Width="fixed"
+        bitOrder="mostSignificantBitFirst" byteOrder="bigEndian"/>
+    </xs:appinfo>
+  </xs:annotation>
```
If we do the nitf-core-types.dfdl.xsd thing, we might also want to define a format that extensions can use so they don't need to copy paste this whole blob.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -853,50 +833,72 @@ The message root is 'NITF'.
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+    </xs:choice>
```
Similar to the header info, would it be possible to avoid the ClassificationInfoset2.0/2.1 elements, and just have some elements be optional or runtime compute lengths? Or maybe the choice can just be a choice of sequence's instead of a choice of elements? That might be easier, but I think you'd need to move SecurityClassification outside of the choice to avoid UPA violation.

jadams-tresys replied:
This was one of the things I was hoping to discuss during the code review.  I knew this 2.0 vs 2.1 stuff was kind of ugly, but I wasn't sure what the best way to cleanly handle it would be.  I'll look into the choice of sequences first.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -149,6 +158,14 @@ The message root is 'NITF'.
+  <xs:complexType name="hexBinaryOrBlob">
+    <xs:choice dfdl:choiceDispatchKey="{ if ($nitf:payloadSizeInBytes gt $nitf:maxBinarySizeInBytes) then 'blob' else 'hb' }">
+      <xs:element name="hexBinaryPayload" dfdl:choiceBranchKey="hb" type="xs:hexBinary" dfdl:lengthKind="explicit" dfdl:length="{ $nitf:payloadSizeInBytes }" />
+      <xs:element name="blobPayload" dfdl:choiceBranchKey="blob" type="xs:anyURI" dfdlx:objectKind="bytes" dfdl:lengthKind="explicit" dfdl:length="{ $nitf:payloadSizeInBytes }" />
```
For consistency, start these with a capital letter?

jadams-tresys commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -125,9 +131,12 @@ The message root is 'NITF'.
+  <xs:complexType name="dateTime">
+    <xs:choice>
+      <xs:element name="dateTime2.1" type="xs:dateTime" dfdl:lengthKind="explicit" dfdl:length="14" dfdl:calendarPatternKind="explicit" dfdl:calendarPattern="yyyyMMddHHmmss" />
+      <xs:element name="dateTime_str" type="xs:string" dfdl:lengthKind="explicit" dfdl:length="14" />
```
I'll look into the choice dispatch.  I think that will work.  Originally I had a separate pattern for the NITF 2.0 stuff as well as a string because of discrepancies in the test data, but if we are just choosing between 2.1 and the string form, the dispatch should work.

jadams-tresys commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -165,36 +182,58 @@ The message root is 'NITF'.
+              <xs:element name="FileCopyNumber" type="nitf:BCS-NP" dfdl:length="5">
+                <xs:annotation>
+                  <xs:appinfo source="http://www.ogf.org/dfdl/">
+                    <dfdl:setVariable ref="nitf:isNITF20" value="{ if (fn:exists(../ClassificationInfo2.0)) then 'true' else 'false' }" />
```
I will make the change.  For some reason I was thinking that we would need to check to make sure ClassificationInfo2.0 existed first, but really I should have just defaulted it to false and set it to true when we parse a 2.0.

jadams-tresys commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -45,17 +45,23 @@ The message root is 'NITF'.
   <xs:annotation>
     <xs:appinfo source="http://www.ogf.org/dfdl/">
 
+      <dfdl:defineVariable name="isNITF20" type="xs:string" defaultValue="false" />
```
Agreed, will change.

jadams-tresys commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -165,36 +182,58 @@ The message root is 'NITF'.
                     <dfdl:assert message="Standard Type must equal 'BF01'" test="{ . eq 'BF01' }" />
                   </xs:appinfo>
-                </xs:annotation>
+                </xs:annotation>-->
```
I was unable to find a documented list of accepted StandardTypes, so I think we will just stick with a 4 character string for now.

jadams-tresys commented:
I believe I have addressed all the comments and have split the project into 2 different schema projects.  To see the new dfdl-nitf-core project, check out https://github.com/jadams-tresys/dfdl-nitf-core

stevedlawrence commented on build.sbt:
```diff
@@ -7,8 +7,9 @@ version := "0.0.1"
 scalaVersion := "2.12.11"
  
 libraryDependencies ++= Seq(
+  "com.tresys" % "dfdl-nitf-core" % "0.0.1",
```
Thoughts on including both schemas in this same repo? That's likely what people that whant just plain nitf would expext. I'm thinking the directory structure might look something like this:

```
.
|-- README
|-- build.sbt
|-- nitf-core
|   |-- build.sbt
|    `-- src/main/resources/nitf.dfdl.xsd
`-- nitf
    |-- build.sbt
    `-- src/main/resources/com/tresys/nitf/xsd/nitf_extension_types.dfdl.xsd
```

So the nitf-core and nitf subdirectories can have a build.sbt in them that allow users to build or publish them individually if they run sbt from those directories. And the root can have a build.sbt that creates subprojects using those subdirectories, which allows easily running "sbt test" from the root to test everything. The subdirectory build.sbt files would look basically as they are now. And I think the root build.sbt would look something like this, so nothing too complicated:

```scala
lazy val root =     Project("root", file("."))
                      .aggregate(nitf, nitfCore)

lazy val nitf =     Project("nitf", file("nitf"))
                      .dependsOn(nitfCore)

lazy val nitfCore = Project("nitf-core", file("nitf-core"))
```

stevedlawrence replied:
The downside to that is if someone wants to parse plain nitf files, all the necessary files aren't in the jar. The nitf schema isn't complete without the `nitf_extension_types.dfdl.xsd` file on the classpath and providing the necessary types. So if it were moved to src/test/resoruces and someone wanted to parse plain nitf, they would need to pull out the file from src/test and put it on the classpath.

I was thinking of this sort of like `nitf-core` is an abstract class that requires some types/elements be defined. The `nitf` projects defines those types, but you could optionally defined those types differenting in the `raster` project.

And so nothing uses nitf-core alone, it's always a dependency to something that "extends" it by implementing those types.

mbeckerle replied:
Is there such a thing as "plain NITF files" for real use, i.e., not just for test?

To my knowledge, there is always at least 1 image data type included. But I might be wrong. 

The "plain NITF file" without even JPEG in it, exists so we can do testing of just that aspect in isolation.

In my mind it's exactly analogous to NACT headers. They're never used alone. We can test them alone, but real usage always adds a payload. 

In the mil-std-2045 schema, we have a schema file, which must be found on the classpath, under src/test/resources, which defines the payload. NACT is similar. 

I guess this boils down to is NITF usable standalone, or are there always extensions?

mbeckerle replied:
There's another technical subtlety here. If I include nitf_extension_types.dfdl.xsd in src/main/resources, then to create a schema for raster I have to control the order of Jars on the classpath so that the raster jar which redefines this file is earlier on the classpath. 

I'm not sure how one controls that classpath order in sbt, maven, etc.  I never thought of the library dependencies in sbt as being an ordered list, but perhaps it is?

stevedlawrence replied:
I think of "plain NITF" as [MIL-STD-2500C](https://gwg.nga.mil/ntb/baseline/docs/2500c/index.html), which does list data types that can be in the payload, e.g. JPEG and JPEG2000. That might not be correct though, and I'm not totally clear how it relates to Raster--I think Raster is just NITF with defined extensions?

stevedlawrence replied:
> There's another technical subtlety here. If I include nitf_extension_types.dfdl.xsd in src/main/resources, then to create a schema for raster I have to control the order of Jars on the classpath so that the raster jar which redefines this file is earlier on the classpath.

That's why I suggest splitting into nitf-core and nitf. nitf-core does not include nitf_extension_types.dfdl.xsd, but only imports it. Then we would have separate nitif and raster projects, each of which would depend on nitf-core and include a different nitf_extension_types.dfdl.xsd file. So there's only one extension types file on the classpath.

One issue with this approach is if you want to parse both NITF and Raster in the same process. Then you would have an issue where the extension types is on the classpath twice. But this seems like a limitation of XML Schema. This like overloading doesn't really seem possible.

mbeckerle replied:
Then nitf-core does not support 'sbt test' because a required include/import file is missing?

The invariant that 'sbt test' always works and tests the module/assembly, that is super important.

stevedlawrence replied:
nitf-core could have its own implementation of the extension types in nitf-core/src/test that is only used for testing--the extension types could just be hexBinary blobs or something simple. Could also mean we could have separate projects like NITFwithJPEG, NITFwithJPEG2000, NITFwithEverything or something if we wanted. And they all could implement different extensions/payloads.

Or maybe another approach is needed. We really need someway to allow for raster to change parts of the nitf schema Splitting into separate nitf-core, nitf, and raster projects seemed like the only way to do it to me, but maybe there's an alternative.

mbeckerle replied:
I found that the order of classpath is in fact the dependency order, that is if this stack-overflow response is believable:

https://stackoverflow.com/questions/33405025/sbt-modify-the-order-of-dependencies-in-the-classpath

So that would mean a schema can include things in separate files, and a jar earlier on the class path can override individual files to provide alternate definitions. 

I would suggest that in build.sbt when we rely on classpath order explicitly, we should say so in comments like
`// override nitf_extension_types.dfdl.xsd file`

stevedlawrence replied:
Ah, that's interesting. So if that's that case, we could keep nitf.dfdl.xsd exactly as it with only a single projet, and in order to do raster we just ensure the raster jar is on the classpath first. I'm a bit concerned that that's a bit fragile and could be easily messed up, since I don't think most people pay attention to classpath order, or even know about the sbt dependency order (I know I didn't).

mbeckerle replied:
I agree that it's not like we have "protected" and "override" keywords to mark our intent, but I don't think this is conceptually too bad. Maybe this is what the XSD redefine element was supposed to be for. Alas that's not part of DFDL, only import and include.  

The other examples we have of this envelope-payload idiom aren't really overriding. E.g., the stanag-5516-with-nact schema depends on nact and stanag-5516 schemas, but the nact schema's own "payload" definition lives in src/test, so it's not really an override by the combined assembly schema, it's supplying a definition that would be missing. 

I suggest we try nitf as one nitf schema, which has the core and a nitf_extension_types.dfdl.xsd in src/master which  requires a dependency on JPEG, as we always have. Nitf's local tests can then still use jpeg images. 

Then the schema for raster (really nitf-with-raster) can depend on nitf and redefine/override the nitf_extension_types.dfdl.xsd file. 

Users of raster will have to insure they have the raster jar and the nitf jar, and the dependencies are in that order. I think that would happen naturally if someone expresses an application dependency on raster only, and the nitf dependency is found transitively.

jadams-tresys replied:
You guys may have mentioned this earlier, but the whole reason for the split was because I couldn't run daffodil from the command line to process Raster data.  Things worked just fine when testing through sbt, but I couldn't figure out a way to reliably process things on the command line.

stevedlawrence replied:
In theory, you should be able to do something like this:
```bash
export DAFFODIL_CLASSPATH="/path/to/raster.jar:/path/to/nitf.jar"
daffodil parse -s com/tresys/nitf/nitf.dfdl.xsd foo.nitf
```
So the raster jar must come first, followed by the nitf dependency. Then you use the classpath resource path found inside the nitf.jar as the root schema. When Daffodil compiles the schema and imports the extensions file, it should find the raster extension file first and use that.

mbeckerle commented on src/main/resources/com/tresys/nitf/xsd/nitf_extension_types.dfdl.xsd:
```diff
@@ -0,0 +1,148 @@
+    </xs:sequence>
+  </xs:complexType>
+
+  <xs:complexType name="imageData">
```
Can some parts of this be made into reusable groups that are defined in the main nitf schema?

Ex: can the entire complex type of MaskTable be factored out as MaskTableType? 

The Choice below at line 132 I can see has to be in this file. But the rest doesn't seem to depend on details within the choice. 

If I'm wrong and this MaskTable does depend on the JPEG behavior, then I'd still suggest pulling out the type as a named complex type just so that type can be named JPEGMaskTable, or MaskTableForJPEGImageData or something like that to make clear it is specific.

stevedlawrence replied:
Makes sense, if some of this is the same as raster, it could be pulled out into the nitif common and shared between the two.

jadams-tresys replied:
I have moved it to its own type now.  There is some overlap with the MaskTable that is defined in Raster, but they call certain elements different names in Raster and other parts are completely different.

mbeckerle commented on src/main/resources/com/tresys/nitf/xsd/nitf_extension_types.dfdl.xsd:
```diff
@@ -0,0 +1,148 @@
+Version 0.0.1  nitf_extension_types.dfdl.xsd
+
+This DFDL schema provides basic, default support for various extendable types
+in NITF, such as TaggedRecordExtensions and ImageData formats.
```
We should make this comment bigger and more tutorial that this is how you add image types into NITF. 

Or better, we can avoid repeating ourselves by putting the instructions in the README.md for NITF, and then we can just reference that from here.

mbeckerle commented on build.sbt:
```diff
@@ -7,8 +7,9 @@ version := "0.0.1"
  
 libraryDependencies ++= Seq(
+  "com.tresys" % "dfdl-nitf-core" % "0.0.1",
   "com.mitre" % "dfdl-jpeg" % "0.0.1",
```
If the jpeg stuff goes under /src/test, then this dependency can be test-only by adding the % "test" at end.

stevedlawrence replied:
Looks like src/main schemas still reference the JFIF schema. Do we want to change that, so the default schema just treats al payloads as blobs, or keep the behavior of using the JPEG schema for JPEG payloads? Now a user can easily override the payload if they want jpeg support, and are src/test could add an overrid for that.

mbeckerle reviewed (changes requested):
Before approving I'd like to see if we can avoid the split into two schemas.

stevedlawrence commented on src/main/resources/com/tresys/nitf/xsd/nitf_extension_types.dfdl.xsd:
```diff
@@ -0,0 +1,148 @@
+    </xs:appinfo>
+  </xs:annotation>
+
+  <xs:complexType name="taggedRecordExtension">
```
I think this is one of those types that must be "implemented". Raster implements it one way, plain NITF implements it another (this way)

mbeckerle commented:
Any ideas on how we can make TravisCI work? This automated test failed because this schema depends on JPEG for testing, and it can't find jpeg. 

Inter-schema dependencies are going to be common.  Should we start publishing DFDL schemas to maven central or other such free internet hosting location (bintray is gone now) so they can be used as dependencies?

stevedlawrence commented:
> Any ideas on how we can make TravisCI work? This automated test failed because this schema depends on JPEG for testing, and it can't find jpeg.

One option would be to just remove the JPEG dependency and always parse payloads as hexBinary. If a user wants to support JPEG, then they can provide their own extension_types.dfdl.xsd file and override the imageData type to include it.

Another option would be to change the github action to checkout the jpeg schema, run publishLocal, and then run the NITF tests.

If we do start publishing schemas somewhere, we'll need to formalize and get better at bumping versions I feel like there's plenty of schemas that have been updated without updating the version number. We also might want to consider things like organizations. If these get published to maven central, maybe we want a convention that the organization for dfdl schemas should be "org.dfdl" or something like that, so that people can easily find all the schemas? Also, since these aren't really standard java things, I wonder if it doesn't really belong in maven central, but some sort of spinoff that's DFDL specific.

mbeckerle commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -865,38 +725,53 @@ The message root is 'NITF'.
+          minOccurs="0" dfdl:occursCountKind="expression" dfdl:occursCount="{ if (fn:nilled(../Downgrade)) then 0 else if (../Downgrade eq '999998') then 1 else 0 }" />-->
+          <!-- This choice shouldn't be necessary, but having an optional element
+               at the end of a choice branch triggers an unparse error see
+               DAFFODIL-XXXX -->
```
Replace XXXX with the ticket number, which is 2562

jadams-tresys replied:
Oops, forgot to do that before I pushed it up.  Will fix

mbeckerle commented on src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd:
```diff
@@ -865,38 +725,53 @@ The message root is 'NITF'.
+          <!-- This choice shouldn't be necessary, but having an optional element
+               at the end of a choice branch triggers an unparse error see
+               DAFFODIL-XXXX -->
+          <xs:choice dfdl:choiceDispatchKey="{ if (fn:nilled(Downgrade)) then 'missing' else if (Downgrade eq '999998') then 'present' else 'missing' }">
```
This is a better solution than using a choice that has backtracking and a discriminator.

mbeckerle reviewed (approved):
+1. Just fix the DAFFODIL-XXXX to the bug number.

stevedlawrence commented on build.sbt:
```diff
@@ -8,7 +8,7 @@ scalaVersion := "2.12.11"
 libraryDependencies ++= Seq(
   "com.mitre" % "dfdl-jpeg" % "0.0.1",
-  "org.apache.daffodil" %% "daffodil-tdml-processor" % "2.6.0" % "test",
+  "org.apache.daffodil" %% "daffodil-tdml-processor" % "3.2.0-SNAPSHOT" % "test",
```
Just curios, does this actually depend on on features in 3.2.0? We might want to only bump this to 3.1.0 so that non daffodil devs can easily test this.

jadams-tresys replied:
Right, this doesn't need 3.2.0-SNAPSHOT.  Will change to 3.1.0.

stevedlawrence reviewed (approved):
+1, just a question about 3.2.0 dependency
